src/main/scala/com/microsoft/hyperspace/index/rules/HyperspaceFilter.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+import com.microsoft.hyperspace.ActiveSparkSession
+import com.microsoft.hyperspace.index.IndexLogEntry
+
+trait HyperspaceFilter {
+
+  /**
+   * @return Failure reason for filtered out indexes.
+   */
+  def reason: String
```

Comment:
I am not still sure about this abstraction since this should be called after `apply` - which is defined in a trait that extends this - is called, and make the implementation have a mutable state.

How about the following (pseudo code)?
```scala
trait IndexFilter[A] {
  def apply(plan: LogicalPlan, entries: Seq[A]): (Seq[A], Option[String]) // Option[String] is the reason.
}

trait SourceFilter extends IndexFilter[IndexLogEntry]
trait PlanFilter extends IndexFilter[(LogicalPlan, Seq[IndexLogEntry])] // I don't think `Map` is required.
```

Reply:
But `reason` will be only used in whyNot API and I don't think it's proper that the return type has the reason string.
The reason string can be empty or non-empty depending on the target index of whyNot API. 

Could you explain more about `the implementation have a mutable state`?

Reply:
Meaning it will need to have `var`.

With the current approach:
```scala
val myFilter = MyFilter(...)
println(myFilter.reason) // empty
val filtered1 = myFilter.apply(...))
val filtered2 = myFilter.apply(...))
println(myFilter.reason) // hopefully the reason is from the latest apply, implementor has to make sure to "clear" the reason.
```

Suggested approach if you don't think the reason string is not "proper" in the return type:
```scala
val myFilter = MyFilter(...)
myFilter.filter(...)
val (filtered1, reason1) = myFilter.filterWithReason(...) // no confusion here
val (filtered2, reason2) = myFilter.filterWithReason(...)
```

Reply:
Oh, seems I didn't share impl. of whyNot API in my mind. It looks like:
```scala
val filteredIndexes = MyFilter(..)
if (!filteredIndexes.contains(targetIndex)) {
  println(MyFilter.reason)
}
...
```

Actually it's `val reason` rather than `def reason`.. and maybe we can do the following if you prefer this structure:

```scala 
def apply(targetIndex, plan, entries): (Seq[A], Option[String]) = {
  val filtered = apply(plan, entries)
  val reasonStr = if (filtered.contains(targetIndex)) { None }
    else { Some(reason) }
  (filtered, reasonStr)
}
```

Reply:
> Actually it's `val reason` rather than `def reason`

How can you do this? Don't you need to update the `MyFilter.reason` whenever you call `apply` since the reason can be different?





> maybe we can do the following if you prefer this structure:

Maybe this will be better because I think a reason should map to "one target" (I think I brought this up in a different thread). Again, we can have a separate `def` that returns a reason, but I am not sure about having `def reason` as an API due to the concerns I shared above (check the comment in my example above).

Reply:
The reason string will be always same for each filter. Otherwise it will ruin readability of filter code by many of if-not-throw(or set msg) - I tried but gave up to support different reasons because of that.

Reply:
Ah OK, then it makes sense. I thought we were doing "many of if-not-throw(or set msg)"

Reply:
Actually, my first impression was the same as @imback82. Shouldn't the message different for each invocation? I'm not sure how setting messages for different reasons can *ruin* the readability that much, and moreover, is it really more important than the user experience? For example, with the current implementation, users are forced to examine their queries to find simple mistakes such as missing columns, and this could be time consuming depending in the size of the query and the tables. This is just an example and this can be more serious - maybe the current approach is sufficient for the current filters, but there could be filters in the future for which a generic class-wide error message is near useless.

Reply:
https://github.com/microsoft/hyperspace/pull/403/files#diff-c6cbe018cfa141e77ee664c5a8fc680f1bd2d6827ca16c2368e34c812b2c7ab0R27 
This commented out code was from that try. 
```def withWhyNotCheck(indexes: Seq[IndexLogEntry], msg: String)```

For each condition & msg inside a filter, call this func with "filtered indexes" and proper msg. 
That was somewhat bothering me as I needed to split one condition into several statements which can be done with one statement & define multiple "filtered" indexes only for this API call. 
If it's not for whyNot API -disabled whyNotAPI-, it just makes more func calls and to avoid that overhead, we need some duplicate code for the API which is also not acceptable to me..

But I agree that user experience is more important.

Reply:
I think splitting one condition into several can actually benefit the reader of the filter code too, unless it's extremely verbose or trivial.

For the performance, we can make it a macro that sets the message if whyNot API is enabled, or does nothing otherwise. I'm not sure whether it can be done with Scala's macros but we can try if the performance overhead is significant.

Reply:
Thanks @sezruby / @clee704 for the discussion. Let's go with the current approach and evolve it if needed/required.